File: 总题库/743.NetworkDelayTime.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int N, int K) {
        int graph[N + 1][N + 1];
        for (int i = 0; i < N + 1; ++i) {
            for (int j = 0; j < N + 1; ++j) {
                graph[i][j] = INT_MAX;
            }
        }

        for (vector<int> triple : times) {
            graph[triple[0]][triple[1]] = triple[2];
        }
        vector<int> dist(N + 1, INT_MAX);
        vector<bool> seen(N + 1, false);
        dist[K] = 0;
        
        while (1) {
            int curNode = -1;
            int curDist = INT_MAX;
            for (int i = 1; i < N + 1; ++i) {
                if (!seen[i] && dist[i] < curDist) {
                    curNode = i;
                    curDist = dist[i];
                }
            }
            if (curNode == -1) {
                break;
            }
            seen[curNode] = true;
            for (int i = 1; i < N + 1; i++) {
                if (graph[curNode][i] < INT_MAX) {
                    dist[i] = min(dist[i], dist[curNode] + graph[curNode][i]);
                }
            }
        }

        int ans = INT_MIN;
        for (int i = 1; i < dist.size(); ++i){
            if (dist[i] > ans) {
                ans = dist[i];
            }
        }
        return ans == INT_MAX ? -1 : ans;
    }
};

// 优先队列实现
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int N, int K) {
        typedef pair<int, int> PII;
        vector<bool> seen(N + 1, false); // 用来标记一个节点是否被访问过
        vector<int> dist(N + 1, INT_MAX); // 距离起始点的最短距离
        unordered_map<int, vector<PII>> graph; // 使用临接表来保存图
        priority_queue<PII, vector<PII>, greater<PII>> heap; // 小顶堆；维护到起始点的最短距离和点

        for (auto &t: times) {
            // 初始化临接表
            graph[t[0]].push_back({t[2], t[1]});
        }

        heap.push({0, K});
        dist[K] = 0;
        while (!heap.empty()) {
            auto t = heap.top();
            heap.pop();
            int curdist = t.first, curNode = t.second;
            if (seen[curNode]) continue;
            seen[curNode] = true;
            for (auto &p: graph[curNode]) {
                if (dist[p.second] > curdist + p.first) {
                    dist[p.second] = curdist + p.first;
                    heap.push({dist[p.second], p.second});
                }
            }
        }
        int ans = *max_element(dist.begin() + 1, dist.end());
        return ans == INT_MAX ? -1 : ans;
    }
};

// floyd-worshall算法
// Dj只能求出一个指定点到其他的最短路，floyd可以求出任意两点之间的最短路；
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int N, int K) {
        const int INF = 0x3f3f3f3f;
        vector<vector<int>> d(N+1, vector<int>(N+1, INF));
        for (int i = 1; i<=N; i++) d[i][i] = 0;

        for (auto &t: times){
            d[t[0]][t[1]] = min(d[t[0]][t[1]], t[2]);
        }

        for (int k = 1; k<=N; k++){
            for (int i = 1; i<=N; i++){
                for (int j =1; j<=N; j++){
                    d[i][j] = min(d[i][j], d[i][k] + d[k][j]);
                }
            }
        }
        int ans = 0;
        for (int i =1; i<=N; i++){
            ans = max(ans, d[K][i]);
        }
        return ans > INF/2 ? -1: ans;
    }
};

File: 总题库/743.NetworkDelayTime.py
# DFS
class Solution:
    def networkDelayTime(self, times, N, K):
        graph = collections.defaultdict(list)
        for u, v, w in times:
            graph[u].append((w, v))
        
        # dist[node]记录的是信号最早到达node的时间。当我们访问node时，若当前信号
        # 是最早到达该节点的，则我们广播这个信号；
        # 为了加快节点速度，在访问每个节点时，若传递该信号的时间比已有信号到达的时间
        # 长，则我们退出该信号，进行剪枝。
        dist = {node: float('inf') for node in range(1, N + 1)}

        def dfs(node, elapsed):
            # elapsed记录的是一个时间戳，代表当前的时间
            if elapsed >= dist[node]:
                return
            dist[node] = elapsed
            for time, nei in sorted(graph[node]):
                dfs(nei, elapsed + time)
        
        dfs(K, 0)
        ans = max(dist.values())
        return ans if ans < float('inf') else -1

# Dijkstra算法
class Solution:
    def networkDelayTime(self, times, N, K):
        graph = collections.defaultdict(list)
        for u, v, w in times:
            graph[u].append((v, w))
        dist = {node: float('inf') for node in range(1, N + 1)}
        seen = [False] * (N + 1)
        dist[K] = 0

        while True:
            cand_node = -1
            cand_dist = float('inf')
            for i in range(1, N + 1):
                if not seen[i] and dist[i] < cand_dist:
                    cand_dist = dist[i]
                    cand_node = i

            if cand_node < 0:
                # 说明已经遍历完所有的节点了
                break;

            seen[cand_node] = True
            for nei, d in graph[cand_node]:
                dist[nei] = min(dist[nei], dist[cand_node] + d)

        ans = max(dist.values())
        return ans if ans < float('inf') else -1

# 堆实现：
class Solution:
    def networkDelayTime(self, times, N, K):
        graph = collections.defaultdict(list)
        for u, v, w in times:
            graph[u].append((v, w))
        
        pq = [(0, K)]
        dist = {}
        while pq:
            d, node = heapq.heappop(pq)
            if node in dist:
                continue
            dist[node] = d
            for nei, d2 in graph[node]:
                if nei not in dist:
                    heapq.heappush(pq, (d + d2, nei))
        return max(dist.values()) if len(dist) == N else -1